R/tostring.R

```diff
@@ -805,14 +1087,159 @@ wrap_string <- function(str, width, collapse = NULL) {
 }
 
 .go_stri_wrap <- function(str, w) {
+  if (is.na(w) || is.nan(w)) {
+    browser()
+  }
+  if (w < 1) {
+    return(str)
+  }
   stringi::stri_wrap(str,
     width = w,
     normalize = FALSE, # keeps spaces
     simplify = TRUE, # If FALSE makes it a list with str elements
-    indent = 0
+    indent = 0,
+    use_length = FALSE # incase the defaul changes, use actual char widths
   )
 }
 
+#' @rdname wrap_string_ttype
+#' @export
+split_word_ttype <- function(str, width, fontspec, min_ok_chars) {
+  chrs <- strsplit(str, "")[[1]]
+  nctt_chars <- nchar_ttype(chrs, fontspec, raw = TRUE)
+  ok <- which(cumsum(nctt_chars) <= width)
+  if (length(ok) < min_ok_chars || length(chrs) - length(ok) < min_ok_chars) {
+    list(
+      ok = character(),
+      remainder = str
+    )
+  } else {
+    list(
+      ok = substr(str, 1, length(ok)),
+      remainder = substr(str, length(ok) + 1, nchar(str))
+    )
+  }
+}
+
+## need a separate path here because **the number of characters**
+## in each part is no longer going to be constant the way it
+## was for monospace
+## this is much slower but still shouldn't be a bottleneck, if it is we'll
+## have to do something else
+#' wrap string given a Truetype font
+#'
+#' @inheritParams wrap_string
+#' @param min_ok_chars numeric(1). Number of minimum characters which much remain
+#' on either side when a word is split
+#' @param wordbreak_ok logical(1). Should breaking within a word be allowed. If, `FALSE`,
+#' attempts to wrap a string to a width narrower than its widest word will result
+#' in an error.
+#' @return `str`, broken up into a word-wrapped vector
+#' @export
+wrap_string_ttype <- function(str,
+                              width,
+                              fontspec,
+                              collapse = NULL,
+                              min_ok_chars = min(floor(nchar(str) / 2), 4, floor(width / 2)),
+                              wordbreak_ok = TRUE) {
+  newdev <- open_font_dev(fontspec)
+  if (newdev) {
+    on.exit(close_font_dev())
+  }
+
+  rawspls <- strsplit(str, "[[:space:]](?=[^[:space:]])", perl = TRUE)[[1]] # preserve all but one space
+  nctt <- nchar_ttype(rawspls, fontspec, raw = TRUE)
+  pts <- which(cumsum(nctt) <= width)
+  if (length(pts) == length(rawspls)) { ## no splitting needed
+    return(str)
+  } else if (length(pts) == 0) { ## no spaces, all one word, split it and keep going
+    if (wordbreak_ok) {
```

Comment:
This feature is not avalilable at the moment externally right? It is "just in case"? (it makes complete sense to have it imo)

Reply:
its not available at the `toString` or `paginate_*` level currently, but it can be used when, e.g., iteratively finding optimal column widths (which is massively more non-trivial with truetype fonts) by checking if a particular width *would* break within a word or not given a not-already-wrapped mpf.

It also could be added to `toString`/`paginate_*`in the future if we want

Reply:
this makes a lot of sense! propose_colwidths is heavily used and sometimes users add very low columnwidths manually, which can generate issues that are hard to understand (like pagination may break)